### PR TITLE
TD3 Code review

### LINKED ISSUE #17: Review of Existing Code
@araffin has done a great job creating this version of the library but has done it mostly solo, so much of the code has never been reviewed. 

I suggest the other maintainers each take responsibility for reviewing a portion of the code. Rather than doing a traditional code review, since it's already committed I suggest we just make a PR with any changes we think should take place, or raise an issue for non-trivial proposals.

I think the priority is code that is used in multiple algorithms and/or defines the public API and which is new. This includes:
  - [x] `common/base_class.py`
  - [x] `common/distributions.py`
  - [x] `common/policies.py`
  - [x] `common/type_aliases.py`

Next would be the individual algorithms:
  - [x] A2C
  - [x] PPO
  - [ ] SAC
  - [ ] TD3

Also new parts of the documentation could use re-reading/confirming:
  - [x] Documentation

If this sounds like a good idea to others, then perhaps we can each claim a few entries on the above list and then start a PR for it? Also feel free to edit the post to break it up differently or to add files I've missed.

### LINKED ISSUE #331: [Bug] action_noise never got used in HER
### 🐛 Bug
Hi when I was trying to use HER with DDPG, I noticed that the action noise I feed into HER never got used by DDPG. 
Please correct me if I am wrong, if you take a look at the code on 197 at HER.py
```
            rollout = self.collect_rollouts(
                self.env,
                n_episodes=self.n_episodes_rollout,
                n_steps=self.train_freq,
                action_noise=self.action_noise,
                callback=callback,
                learning_starts=self.learning_starts,
                log_interval=log_interval,
            )

```
I have printed out self.action_noise, it is NONE. the action_noise got fed into self.model during the init part (line 103 as kwargs), but I think it never got assigned to it self as an attribute, maybe we shall do self.model.action_noise there? Let me know if I am mistaken.

### PR DESCRIPTION
**Pull request by ernestum**
## Description
Discussion MR for the TD3 implementation

## Motivation and Context

This code review is part of the review of existing code #17.

Also fixes issues with terminal observation.

closes #17
closes #331 

- [ ] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [ ] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [ ] I have reformatted the code using `make format` (**required**)
- [ ] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [ ] I have ensured `make pytest` and `make type` both pass. (**required**)
- [ ] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

File: stable_baselines3/common/policies.py [hunk 1]
```diff
@@ -119,12 +119,11 @@ def extract_features(self, obs: th.Tensor) -> th.Tensor:
         preprocessed_obs = preprocess_obs(obs, self.observation_space, normalize_images=self.normalize_images)
         return self.features_extractor(preprocessed_obs)
 
-    def _get_data(self) -> Dict[str, Any]:
+    def _get_data_to_reconstruct_model(self) -> Dict[str, Any]:
```
Comment (araffin) [inline stable_baselines3/common/policies.py:122]: suggestion: `get_constructor_parameters()` but I'm ok with this name too ;)

**Comment by araffin** [comment]
@ernestum are you done with the edits?

**Comment by ernestum** [comment]
No I still wanted to implement the more convenient way to specify the training frequency based on either steps or episodes. I am on it right now.

I noticed that SAC does not support updating after n episodes (just after n steps). Was it supposed to be that way? Because after the patch that will be possible.

**Comment by ernestum** [comment]
Another question regarding [this line](https://github.com/DLR-RM/stable-baselines3/blob/52003e0bd8ff7b284b96c240bb98e06dd2cc6100/stable_baselines3/td3/td3.py#L163) (and all other lines where a `polyak_update` is done):
```python
if gradient_step % self.policy_delay == 0:
    ...
    polyak_update(...)
```
Note that this triggers a polyak update in each first iteration of the outer `for gradient_step in range(gradient_steps)` loop since `0 % x == 0` for all `x`.
 In the case where we set `train_freq=1` the policy is not delayed since it will be polyak-updated after each step. I guess this should be more like `if self.total_train_steps % self.policy_delay == 0:` right?

**Comment by Miffyli** [comment]
> I noticed that SAC does not support updating after n episodes (just after n steps). Was it supposed to be that way? Because after the patch that will be possible.

Jumping in here to comment on this. Is this something people do on robotics side? I have personally not seen "update every N episodes" type of setups, but I mainly focus on games and non-robotics envs. It sounds very sensitive to problems (e.g. agent gets better -> episodes last longer -> less updates), and support for this could be confusing. If support for this is added it should be blatantly clear how "update every N steps" and "update every N episodes" function together (e.g. do they override each other? In which order? Do they both apply?)

**Comment by araffin** [comment]
>I noticed that SAC does not support updating after n episodes (just after n steps). Was it supposed to be that way? Because after the patch that will be possible.

why not? (I've been using that feature on a real robot...)

> In the case where we set train_freq=1 the policy is not delayed since it will be polyak-updated after each step. I guess this should be more like if self.total_train_steps % self.policy_delay == 0: right?

oh... nice catch ;)

> Is this something people do on robotics side? 

yes, mostly because everything happens in real time, you cannot afford having inconsistent and slow control frequency because your policy is updating (whereas in simulation, you can stop the simulation until you are ready again).

>If support for this is added it should be blatantly clear how "update every N steps" and "update every N episodes" function together (e.g. do they override each other? In which order? Do they both apply?)

they are already implemented, and there is a big warning when one is trying to use both at the same time.
@ernestum proposal was in fact to make things clearer.

File: stable_baselines3/common/off_policy_algorithm.py [hunk 2]
```diff
@@ -83,9 +81,8 @@ def __init__(
         batch_size: int = 256,
         tau: float = 0.005,
         gamma: float = 0.99,
-        train_freq: int = 1,
+        train_freq: Union[int, Tuple[int, str]] = (1, "episode"),
```
Comment (araffin) [inline stable_baselines3/common/off_policy_algorithm.py:84]: please keep the default ("step")

File: stable_baselines3/common/off_policy_algorithm.py [hunk 3]
```diff
@@ -250,11 +239,17 @@ def learn(
         callback.on_training_start(locals(), globals())
 
         while self.num_timesteps < total_timesteps:
+            if isinstance(self.train_freq, int):
```
Comment (araffin) [inline stable_baselines3/common/off_policy_algorithm.py:242]: missing comment

Comment (araffin) [reply stable_baselines3/common/off_policy_algorithm.py:242]: but I would do that transform directly in the constructor, no?

File: stable_baselines3/common/off_policy_algorithm.py (hunk 2 above)
Comment (ernestum) [reply stable_baselines3/common/off_policy_algorithm.py:84]: I was assuming the default should be 1 step? So either `1` or `(1, "step")` should be fine? Why are you suggesting 1 episode? Or am I getting something wrong here?

Comment (araffin) [reply stable_baselines3/common/off_policy_algorithm.py:84]: yes, the default should be one step (but that's not what it is right now, no?) I read `(1, "episode"),`

File: stable_baselines3/common/off_policy_algorithm.py (hunk 3 above)
Comment (ernestum) [reply stable_baselines3/common/off_policy_algorithm.py:242]: You mean I should add a comment describing the translation from the old style to the new style?

I deliberately kept the transition here since otherwise we would 1. need to store `n_steps` and `n_episodes` in the `OffPolicyAlgorithm` and 2. the user could not change the `train_freq` in hindsight or during training. Its unlikely that it is needed but WHEN you need it it is really annoying.

File: stable_baselines3/common/off_policy_algorithm.py [hunk 4]
```diff
@@ -128,23 +125,15 @@ def __init__(
...
-            warnings.warn(
-                "You passed a positive value for `train_freq` and `n_episodes_rollout`."
-                "Please make sure this is intended. "
-                "The agent will collect data by stepping in the environment "
-                "until both conditions are true: "
-                "`number of steps in the env` >= `train_freq` and "
-                "`number of episodes` > `n_episodes_rollout`"
-            )
+        if isinstance(train_freq, tuple) and train_freq[1] not in ("step", "episode"):
+            warnings.warn(f"The units of the train_freq must be either `step` or `episode` not `{train_freq[1]}`!")
```
Comment (araffin) [inline stable_baselines3/common/off_policy_algorithm.py:136]: why not an assert?

File: stable_baselines3/common/off_policy_algorithm.py (hunk 3 above)
Comment (araffin) [reply stable_baselines3/common/off_policy_algorithm.py:242]: >You mean I should add a comment describing the translation from the old style to the new style?

yes

>1. need to store n_steps and n_episodes in the OffPolicyAlgorithm 

this is already the case (`self.n_episodes_rollout` and `self.train_freq`)

> 2. the user could not change the train_freq in hindsight or during training.

it can if we store it.
Also, if needed, the user can direct call `collect_rollout` alone if needed.

Last thing, I think the check can be simplified if we have an assert about `self.train_freq[1]` before (see comment above ;))

Comment (ernestum) [reply stable_baselines3/common/off_policy_algorithm.py:242]: nope `self.n_episodes_rollout` is gone now. only `self.train_freq` is left and it is split up into old-style `n_steps` and `n_episodes` before passing it to `collect_rollouts` since I was too lazy to also change that implementation too (but we should do that in the future).

I did not quite get what you mean with the user calling `collect_rollout` directly.

Good point about normalizing `self.train_freq` to be a tuple in the constructor. I will do that.

File: stable_baselines3/common/off_policy_algorithm.py (hunk 4 above)
Comment (ernestum) [reply stable_baselines3/common/off_policy_algorithm.py:136]: Good point I will change that.

File: stable_baselines3/common/off_policy_algorithm.py (hunk 3 above)
Comment (araffin) [reply stable_baselines3/common/off_policy_algorithm.py:242]: what I meant was keeping `self.n_episodes_rollout` (that would be nice to load old models too) but computing the value of `self.train_freq` and `self.n_episodes_rollout` in the constructor.

>I did not quite get what you mean with the user calling collect_rollout directly.

the same way the user can call `model.train()` (just updating the model), the user can also directly call `model.collect_rollout()` which the parameters the user wants.

File: stable_baselines3/common/off_policy_algorithm.py (hunk 2 above)
Comment (ernestum) [reply stable_baselines3/common/off_policy_algorithm.py]: Fixed it

File: stable_baselines3/common/off_policy_algorithm.py [hunk 5]
```diff
@@ -128,23 +125,17 @@ def __init__(
...
-            warnings.warn(
-                "You passed a positive value for `train_freq` and `n_episodes_rollout`."
-                "Please make sure this is intended. "
-                "The agent will collect data by stepping in the environment "
-                "until both conditions are true: "
-                "`number of steps in the env` >= `train_freq` and "
-                "`number of episodes` > `n_episodes_rollout`"
-            )
+        if isinstance(self.train_freq, int):
+            self.train_freq = (self.train_freq, "step")
```
Comment (araffin) [inline stable_baselines3/common/off_policy_algorithm.py:136]: what do you think of using NamedTuple? so later you can do `train_freq.n` and `train_freq.type` instead of `train_freq[1]` ?

Comment (ernestum) [reply stable_baselines3/common/off_policy_algorithm.py:136]: Sure but can I have `train_freq.quantity` and `train_freq.unit`?

Comment (araffin) [reply stable_baselines3/common/off_policy_algorithm.py:136]: `quantity` sounds a bit weird to me... I think you wrote `frequency` in the docstring, no?

File: stable_baselines3/common/off_policy_algorithm.py (hunk 3 above)
Comment (ernestum) [reply stable_baselines3/common/off_policy_algorithm.py]: I would be reluctant to add the public `self.n_episodes_rollout` back. This way we basically build an API with two ways to specify the number of epochs between two training steps and introduce back the issue that the value of `self.n_episodes_rollout` can conflict with `self.train_freq`. If you like I can come up with a separate solution to ensure backwards compatibility with old stored models.

Since, as you pointed out, `model.collect_rollout()` is part of the public API and it still has the old-style `n_episodes` and `n_steps` interface that one should actually be updated to the new tuple format as well and I should stop being lazy!

Comment (araffin) [reply stable_baselines3/common/off_policy_algorithm.py]: >I would be reluctant to add the public self.n_episodes_rollout back

Fair enough ;)

>nterface that one should actually be updated to the new tuple format as well and I should stop being lazy!

if you do so, you may also need to remove the `replay_buffer` parameter ;) (which was here for historic reasons)

**Comment by araffin** [comment]
The current failure on the CI is due to a newer version of numpy, we should replace `np.bool` by `bool` and that should solve it ;)

File: stable_baselines3/common/off_policy_algorithm.py (hunk 5 above)
Comment (ernestum) [reply stable_baselines3/common/off_policy_algorithm.py:136]: You are right. I was borrowing terminology from [here](https://en.wikipedia.org/wiki/Unit_of_measurement) and messed it up. In our case the `train_freq` is the quantity, your `n` is the magnitude unit of measurement must be either `step` or `episode` (but this could be extended in the future). So I propose `train_freq.magnitude` and `train_freq.unit`.

**Comment by ernestum** [comment]
> The current failure on the CI is due to a newer version of numpy, we should replace `np.bool` by `bool` and that should solve it ;)

Thanks for the hint I will look into fixing it tomorrow.

File: stable_baselines3/common/off_policy_algorithm.py (hunk 3 above)
Comment (ernestum) [reply stable_baselines3/common/off_policy_algorithm.py]: noted ...

File: stable_baselines3/common/off_policy_algorithm.py (hunk 5 above)
Comment (araffin) [reply stable_baselines3/common/off_policy_algorithm.py:136]: I would really for go `train_freq.frequency` or `train_freq.interval` (so discrete values, that's why I proposed the `n` at first, from the maths side) as `magnitude` sounds like a continuous value to me.
and yes for the `train_freq.unit` ;)

**Comment by araffin** [comment]
@AdamGleave my last commit (https://github.com/DLR-RM/stable-baselines3/pull/245/commits/d83cdc9687b57b70ab1559fb398b3833e7bc0d2b) might affect your imitation learning lib ;) (it's about having the right terminals)

**Review by araffin — APPROVED** [review]
LGTM, thanks for the fixes and the discussion =)

**Comment by AdamGleave** [comment]
> @AdamGleave my last commit ([d83cdc9](https://github.com/DLR-RM/stable-baselines3/commit/d83cdc9687b57b70ab1559fb398b3833e7bc0d2b)) might affect your imitation learning lib ;) (it's about having the right terminals)

Thanks for flagging! Tagging @shwang FYI: we have previously been seeing unnormalized terminal observations, this PR will fix that.
